### DISCUSSION TITLE
Delete User from Application - returns 401

### DISCUSSION THREAD

**Discussion by fxmb** [supabase/supabase#5208] [Questions]
If I try to remove a user from the applicaiton I always get a 401 not authorized.
This is my code I am calling from `api/` in Nextjs. 
I do not understand what I am missing.

Any help is highly appreciated :-) 

```
export default async function deleteUser(req, res) {
    const { userId } = req.body;
    const { data, user, error } = await supabase.auth.api.deleteUser(
        userId,
        process.env.NEXT_PUBLIC_SUPABASE_SERVICE_KEY
    );
    // Send a 400 response if something went wrong
    if (error) return { error: error.message };
    // Send 200 success if there were no errors!
    // and also return a copy of the object we received from Supabase
    return res.status(200).json({ user, data });
}
```

**Comment by fanclubfm** [comment]
Create your supabase client like this (just my choice of preference to call it supabaseAdmin, you can call it supabase if you want)
import { createClient } from "@supabase/supabase-js"

export const supabaseAdmin = createClient(process.env.NEXT_PUBLIC_SUPABASE_URL, process.env.SUPABASE_SERVICE_KEY)

then in your api function
const { data, user, error } = await supabaseAdmin.auth.api.deleteUser(userId);

**Reply by fanclubfm** [reply]
Also - don't do this process.env.NEXT_PUBLIC_SUPABASE_SERVICE_KEY !!  change it to SUPABASE_SERVICE_KEY in your env file.   Otherwise you will expose it to the frontend when you deploy.

**Reply by fxmb** [reply]
@fanclubfm thank you for your solution! It actually works. However, I cannot remove the prefix NEXT_PUBLIC - it will fail to read the key then and throws a `Error: supabaseKey is required.`. Guess that is because we are calling from frontend and it isnt exposed anymore?

**Reply by undrtherug** [reply]
Thank you for this. Much appreciated

**Comment by j314h** [comment]
hello, I have the same problem and I tried all the means and I always have the error: 'User not allowed' of status: 401 have you managed to find a solution?

**Reply by fanclubfm** [reply]
Can you share code example how you are using it?  I am using it like this and it just works.

```
import { supabaseAdmin } from "@/lib/supabase-admin"

const handler = async (req, res) => {
  const { user_id } = req.query
  if (req.method === "GET") {
    const { data: user, error } = await supabaseAdmin.auth.api.getUserById(user_id)
    return res.send({ user, error })
  } else if (req.method === "DELETE") {
    const { data, user, error } = await supabaseAdmin.auth.api.deleteUser(user_id)
    return res.send({ user, data, error})
  }
  return res.send("Method not allowed")
}

export default handler

```

And my supabaseAdmin is just this (note I am using service role key)
```
import { createClient } from "@supabase/supabase-js"

export const supabaseAdmin = createClient(process.env.NEXT_PUBLIC_SUPABASE_URL, process.env.SUPABASE_SERVICE_ROLE_KEY)
```

**Reply by j314h** [reply]
OK, I found a solution, I don't know if it's a good solution, but it does the job.

First of all I would like to say that the documentation here
https://supabase.com/docs/reference/javascript/auth-api-deleteuser
 is not really good. Indeed we can see that the function presented, contains two arguments while in reality, it only takes one argument otherwise vscode gives you an error

then the problem of understanding the documentation is that we are not told that we will have to modify or create a new supabase client configuration in our project.
let me explain :

when you start a front project you initialize a supabase client like this:
```js
CUSTOMER API KEY
const SUPABASE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoiYW5vbiIsImlhdCI6MTY0MzY0MTI0MiwiZXhwIjoxOTU5MjE3MjQyfQ.C4pkMFCiXtYMjGJ3YjsCG7KBpS_g3tkrFG'
EXAMPLE USE
const SUPABASE_URL = "https://bharrubfpjkhqdxadxef.supabase.co"

const supabase = createClient(SUPABASE_URL, process.env.SUPABASE_KEY);
```
SUPABASE_KEY here is anon key

except that with this client the function
supabase.auth.api.deleteUser(idUser) will not work because there will be no rights

so i searched for a way to modify this key in this client at the time of using deleteUser() but i did not succeed

so I created a second client to run this function here is my code

```js
const userId = supabase.auth.user().id;

  const supa = createClient(import.meta.env.VITE_URL_SUPABASE, import.meta.env.VITE_SECRET);

  const { error } = await supa.auth.api.deleteUser(userId);
```
I specify that I am on the back side, and that import.meta.env.VITE_SECRET is the `service_role`.

as I said it works but I don't know at all if it's the right solution, I would prefer to modify my client just long enough to execute the deleteUser() function but I haven't found how to do it.

I think that the doc must be revised, and see to give the possibility of modifying the already existing client in order to switch between the `anon` key and the `service_role`

anyway thank you for your answer
